File: FindMinRotatedSortedArr.java
/**
Time Complexity - O(N) where N is the size of nums.
Space Complexity - O(1).
 */
class Solution {
    public int findMin(int[] nums) {
        int left = 0, right = nums.length - 1;
        
        if(nums.length == 1 || nums[left] < nums[right])
            return nums[0];
        
        while(left <= right)
        {     
            int mid = (right - left) / 2 + left;
            
            if(mid > 0 && nums[mid] < nums[mid - 1])
                return nums[mid];
            
            if(mid < nums.length - 1 && nums[mid + 1] < nums[mid])
                return nums[mid + 1];
            
            if(nums[right] >= nums[mid]) 
                //if mid is not the smallest element, 
                //then if the nums[right] is > nums[mid] we can confirm that 
                //the min element does not exist to the right of the array.
                right = mid - 1;
            else
                left = mid + 1;
            
        }
        return -1;
    }
}

File: findMinRotatedSortedArr.cpp
// Time Complexity : O(log n) in worst and avg case, O(1) in best case i.e. when it is sorted in increasing order
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0;
        int right = nums.size()-1;
        while(left<=right)
        {
            if(nums[left]<=nums[right])
            {
                return nums[left];
            }
            int mid = left+(right-left)/2;

            if(nums[mid]<nums[mid+1] && nums[mid]<nums[mid-1])
            {
                return nums[mid];
            }
            else if((nums[mid]<nums[right]))
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }

        }
        return 123;
    }
};